doc/guides/porting/arch.rst

```diff
@@ -478,38 +568,189 @@ To enable the :option:`CONFIG_HW_STACK_PROTECTION` feature, the system must
+  the guard is unnecessary and shouldn't be unconditionally reserved.
+  This is the case when privilege elevation stacks are not inside the stack
+  object.
+
+* The required guard size is variable and depends on context. For example, some
+  ARM CPUs have lazy floating point stacking during exceptions and may
+  decrement the stack pointer by a large value without writing anything,
+  completely overshooting a minimally-sized guard and corrupting adjacent
+  memory. Rather than unconditionally reserving a larger guard, the extra
+  memory is carved out if the thread uses floating point.
+
+User mode enabled
 =================
 
-If the target platform enables paging and requires drivers to memory-map
-their I/O regions, :option:`CONFIG_MMU` needs to be enabled and the
-:cpp:func:`arch_mem_map()` API implemented.
+Enabling user mode activates two new requirements:
+
+* A separate fixed-sized privilege mode stack, specified by
+  :option:`CONFIG_PRIVILEGED_STACK_SIZE`, must be allocated that the user
+  thread cannot access. It is used as the stack by the kernel when handling
+  system calls. If stack guards are implemented, a stack guard region must
+  be able to be placed before it, with support for carve-outs if necessary.
+
+* The memory protection hardware must be able to program a region that exactly
+  covers the thread's stack buffer, tracked in ``thread.stack_info``. This
+  implies that :c:macro:`ARCH_THREAD_STACK_SIZE_ADJUST()` will need to round
+  up the requested stack size so that a region may cover it, and that
+  :c:macro:`ARCH_THREAD_STACK_OBJ_ALIGN()` is also specified per the
+  granularity of the memory protection hardware.
+
+This become more complicated if the memory protection hardware requires that
+all memory regions be sized to a power of two, and aligned to their own size.
+This is common on older MPUs and is known with
+:option:`CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT`.
+
+``thread.stack_info`` always tracks the user-accessible part of the stack
+object, it must always be correct to program a memory protection region with
+user access using the range stored within.
+
+Non power-of-two memory region requirements
+-------------------------------------------
+
+On systems without power-of-two region requirements, the reserved memory area
+for threads stacks defined by :c:macro:`K_THREAD_STACK_RESERVED` may be used to
+contain the privilege mode stack. The layout could be something like:
+
+::
+
+   +------------------------------+ <- thread.stack_obj
+   | Other platform data          |
+   +------------------------------+
+   | Guard region (if enabled)    |
+   +------------------------------+
+   | Guard carve-out (if needed)  |
+   |..............................|
+   | Privilege elevation stack    |
+   +------------------------------| <- thread.stack_obj +
+   | Stack buffer                 |      K_THREAD_STACK_RESERVED =
+   .                              .      thread.stack_info.start
+
```

Comment:
@andrewboie @ioannisg 

Is this diagram correct for Arm platforms? In the discussion thread Andrew commented "It's simpler this way since ARM is unconditionally generating separate privilege stacks". And from what I can see the Arm thread privilege stack is coming from a separate object (haven't quite wrapped my head around that).

Reply:
As I said earlier: The doc is a guide to the best known method for configuring this stuff for the different memory management hardware capabilities.

Arm isn't currently following the best known method. It generates separate privilege mode stacks for all ARM platforms.

I do not want to change the diagrams in the porting guide to reflect the current state of ARM as that would give people doing new arch ports the wrong idea.

Reply:
This is the bug link for this ARM issue: https://github.com/zephyrproject-rtos/zephyr/issues/24048

Reply:
Understand. Thank you for the clarification (and confirmation).